File: backend/services/workflow_validation_service.py
"""
Workflow validation service (SRP).
Extracts validation logic from debug routes.
"""
from typing import Dict, Any, List


def validate_workflow_definition(
    workflow_id: str,
    definition: dict,
) -> Dict[str, Any]:
    """
    Validate a workflow definition for potential issues.
    Returns dict with workflow_id, valid, issues, warnings, node_count, edge_count.
    """
    nodes = definition.get("nodes") or []
    edges = definition.get("edges") or []
    if not isinstance(nodes, list):
        nodes = []
    if not isinstance(edges, list):
        edges = []

    issues: List[Dict[str, Any]] = []
    warnings: List[Dict[str, Any]] = []

    # Check for nodes without edges (defensive: node.get("id") for malformed nodes)
    node_ids = set()
    for i, node in enumerate(nodes):
        if isinstance(node, dict) and "id" in node:
            node_ids.add(node["id"])
        elif isinstance(node, dict):
            node_ids.add(f"unknown-{i}")
    connected_nodes = set()
    for edge in edges:
        if isinstance(edge, dict):
            if "source" in edge:
                connected_nodes.add(edge["source"])
            if "target" in edge:
                connected_nodes.add(edge["target"])

    orphan_nodes = node_ids - connected_nodes
    if orphan_nodes:
        warnings.append({
            "type": "orphan_nodes",
            "message": f"Found {len(orphan_nodes)} disconnected nodes",
            "nodes": list(orphan_nodes),
        })

    # Check for missing start/end nodes
    node_types = [
        node.get("type") for node in nodes
        if isinstance(node, dict)
    ]
    if "start" not in node_types:
        issues.append({
            "type": "missing_start",
            "message": "Workflow has no START node",
            "severity": "error",
        })

    if "end" not in node_types:
        warnings.append({
            "type": "missing_end",
            "message": "Workflow has no END node",
            "severity": "warning",
        })

    # Check for cycles
    if _has_cycle(node_ids, edges):
        issues.append({
            "type": "cycle_detected",
            "message": "Workflow contains a cycle (except for loops)",
            "severity": "error",
        })

    # Check agent nodes for configuration
    for node in nodes:
        if not isinstance(node, dict):
            continue
        if node.get("type") == "agent":
            node_data = node.get("data", {})
            agent_config = node_data.get("agent_config", {})

            if not agent_config.get("system_prompt"):
                warnings.append({
                    "type": "missing_system_prompt",
                    "message": f"Agent node '{node.get('id')}' has no system prompt",
                    "node_id": node.get("id"),
                })

    return {
        "workflow_id": workflow_id,
        "valid": len(issues) == 0,
        "issues": issues,
        "warnings": warnings,
        "node_count": len(nodes),
        "edge_count": len(edges),
    }


def _has_cycle(node_ids: set, edges: list) -> bool:
    """Check if workflow graph has a cycle. Skips malformed edges."""
    visited = set()
    rec_stack = set()

    def dfs(node_id: str) -> bool:
        visited.add(node_id)
        rec_stack.add(node_id)

        for edge in edges:
            if not isinstance(edge, dict) or "source" not in edge or "target" not in edge:
                continue
            if edge["source"] == node_id:
                target = edge["target"]
                if target not in visited:
                    if dfs(target):
                        return True
                elif target in rec_stack:
                    return True

        rec_stack.remove(node_id)
        return False

    for nid in node_ids:
        if nid not in visited:
            if dfs(nid):
                return True
    return False

File: backend-java/src/main/java/com/workflow/service/WorkflowValidationService.java
package com.workflow.service;

import com.workflow.dto.Edge;
import com.workflow.dto.Node;
import com.workflow.entity.Workflow;
import com.workflow.repository.WorkflowRepository;
import com.workflow.util.ErrorMessages;
import com.workflow.util.ObjectUtils;
import com.workflow.util.RepositoryUtils;
import com.workflow.util.WorkflowMapper;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * SRP-1: Workflow validation logic extracted from DebugController.
 */
@Service
public class WorkflowValidationService {
    private final WorkflowRepository workflowRepository;
    private final WorkflowMapper workflowMapper;

    public WorkflowValidationService(WorkflowRepository workflowRepository, WorkflowMapper workflowMapper) {
        this.workflowRepository = workflowRepository;
        this.workflowMapper = workflowMapper;
    }

    public Map<String, Object> validate(String workflowId) {
        Workflow w = RepositoryUtils.findByIdOrThrow(workflowRepository, workflowId, ErrorMessages.WORKFLOW_NOT_FOUND);
        Map<String, Object> def = w.getDefinition();
        List<Node> nodes = workflowMapper.extractNodes(def);
        List<Edge> edges = workflowMapper.extractEdges(def);

        List<Map<String, Object>> issues = new ArrayList<>();
        List<Map<String, Object>> warnings = new ArrayList<>();

        Set<String> nodeIds = new HashSet<>();
        Set<String> connected = new HashSet<>();
        for (Node n : nodes) {
            if (n.getId() != null) nodeIds.add(n.getId());
        }
        for (Edge e : edges) {
            if (e.getSource() != null) connected.add(e.getSource());
            if (e.getTarget() != null) connected.add(e.getTarget());
        }
        Set<String> orphans = new HashSet<>(nodeIds);
        orphans.removeAll(connected);
        if (!orphans.isEmpty()) {
            warnings.add(Map.of("type", "orphan_nodes", "message", "Found " + orphans.size() + " disconnected nodes", "nodes", orphans));
        }

        List<String> types = nodes.stream()
                .map(n -> ObjectUtils.safeGet(n.getType(), com.workflow.dto.NodeType::getValue))
                .filter(Objects::nonNull)
                .toList();
        if (!types.contains("start")) {
            issues.add(Map.of("type", "missing_start", "message", "Workflow has no START node", "severity", "error"));
        }
        if (!types.contains("end")) {
            warnings.add(Map.of("type", "missing_end", "message", "Workflow has no END node", "severity", "warning"));
        }

        Map<String, Object> result = new HashMap<>();
        result.put("workflow_id", workflowId);
        result.put("valid", issues.isEmpty());
        result.put("issues", issues);
        result.put("warnings", warnings);
        result.put("node_count", nodes.size());
        result.put("edge_count", edges.size());
        return result;
    }
}
